docs/enhancements/fep-000-api-device-fleet.md

```diff
@@ -0,0 +1,487 @@
+for devices.
+
+The proposed changes include:
+
+* Reincorporating device.spec as a visible part through this frontend API.
+
+* The `device.spec.os.image` contains a reference to the image to be used for the device.
+
+* The `device.spec.templateVersion` field is moved to `device.metadata.resourceVersion` opaque string identifies
+  the version of the device spec, this is used for tracking changes. This is moved from the metadata level, so
+  when we write in the spec field, the contents are then not modified by the server.
+
+* The `device.spec.{containers, systemd}` fields are moved under `device.spec.monitoring.{containers, systemd}`
+
+* The `device.spec.monitoring.containers` contains a list of container patterns to be watched on the device (as in the fleet spec)
+
+* The `device.spec.monitoring.systemd` contains a list of systemd service patterns to be watched on the device (as in the fleet spec)
+
+* The `device.spec.config` contains a list of zero or more config items.
+
+* The config items could be inline definition of config files (`InlineConfigProviderSpec`)
+
+* Some config items reference external configurations (`KubernetesSecretProviderSpec`, `GitConfigProviderSpec`)
+
+* `GitConfigProviderSpec` config types add a `mountPath` field for consistency, in the same way k8s secrets add such field.
+
+* External references (image or config items) must support floating tags (`latest`, `main`, `v1.0`), although
+  **when managed from a fleet, the floating tags will be frozen as hashes** on the device spec by the
+  fleet controller.
+
+* Some external items are actually non-versioned like the K8s secrets, so are inherently floating.
+
+Complete representation of the proposed device:
+
+```yaml
+apiVersion: v1alpha1
+kind: Device
+metadata:
+  name: fab9839018890a88b898b980f8f809f8e8ac333761977d987a777a777a987ccce
+  labels:
+    site: factory-a
+    deviceType: optical-inspector
+    stage: production
+  annotations:
+    fleet-controller/resourceVersion: "optical-inspector-production-fleet-0000001" # the fleet controller uses this annotation to track the latest template applied
+  owner: Fleet/optical-inspector-production-fleet
+  resourceVersion: "0001/hash/implementatio-specific"  # instead of templateVersion, hiding implementation specific details
+spec:
+  os:
+    image: quay.io/redhat/rhde:9.2
+  config:
+    - name: rendered-config
+      configType: InlineConfigProviderSpec
+      inline:
+        ignition:
+          version: 3.4.0
+        storage:
+          files:
+            - contents:
+                source: >-
+                  data:,This%20system%20is%20managed%20by%20flightctl.%0A
+              mode: 422
+              overwrite: true
+              path: "/etc/motd"
+
+    - name: config-from-git
+      configType: GitConfigProviderSpec
+      repository: my-repo
+      ref: v1.0.0
+      path: /path/to/config/in/git
+      mountPath: /path/to/config/in/device
```

Comment:
GitConfigProviderSpec would mean that the agent would clone the repo locally? Or is that taken care of server side and only the file is passed directly via IGN? Just making sure I understand how this is expected to work and if this configType is necessary for agent level spec.

Reply:
Question:
- Do we currently have more than one configType `InlineConfigProviderSpec` that the agent supports?

Reply:
Just as today, the server side merges all of the configs and serves ignition to the agent.

Reply:
So to be clear the only template configType will be InlineConfigProviderSpec and the API's will not change here at all for the device?

Reply:
correct.